aphrodite/endpoints/api_server_kobold.py

```diff
@@ -0,0 +1,196 @@
+# Adapted from openai/api_server.py and tgi-kai-bridge
+
+import argparse
+import asyncio
+import json
+
+from http import HTTPStatus
+from typing import List, Tuple, AsyncGenerator
+
+import uvicorn
+from fastapi import FastAPI, APIRouter, Request
+from fastapi.responses import JSONResponse, StreamingResponse
+
+from aphrodite.engine.args_tools import AsyncEngineArgs
+from aphrodite.engine.async_aphrodite import AsyncAphrodite
+from aphrodite.common.logger import init_logger
+from aphrodite.common.outputs import RequestOutput
+from aphrodite.common.sampling_params import SamplingParams
+from aphrodite.transformers_utils.tokenizer import get_tokenizer
+from aphrodite.common.utils import random_uuid
+from aphrodite.endpoints.protocol import KAIGenerationInputSchema
+
+TIMEOUT_KEEP_ALIVE = 5  # seconds
+
+logger = init_logger(__name__)
+served_model: str = "Read Only"
+engine: AsyncAphrodite = None
+
+app = FastAPI()
+kai_api = APIRouter()
+extra_api = APIRouter()
+
+def create_error_response(status_code: HTTPStatus, message: str) -> JSONResponse:
+    return JSONResponse({"msg": message, "type": "invalid_request_error"},
+                        status_code=status_code.value)
+
+@app.exception_handler(ValueError)
+def validation_exception_handler(request, exc):  # pylint: disable=unused-argument
+    return create_error_response(HTTPStatus.UNPROCESSABLE_ENTITY, str(exc))
+
+def prepare_engine_payload(kai_payload: KAIGenerationInputSchema) -> Tuple[SamplingParams, List[int]]:
+    """Create SamplingParams and truncated input tokens for AsyncEngine"""
+
+    if kai_payload.max_context_length > max_model_len:
+        raise ValueError(
+            f"max_context_length ({kai_payload.max_context_length}) must be less than or equal to "
+            f"max_model_len ({max_model_len})"
+        )
+
+    sampling_params = SamplingParams(max_tokens=kai_payload.max_length)
+
+    # KAI spec: top_k == 0 means disabled, aphrodite: top_k == -1 means disabled
+    # https://github.com/KoboldAI/KoboldAI-Client/wiki/Settings
+    top_k = kai_payload.top_k if kai_payload.top_k != 0.0 else -1
+
+    sampling_params = SamplingParams(
+        n=kai_payload.n,
+        best_of=kai_payload.n,
+        repetition_penalty=kai_payload.rep_pen,
+        temperature=kai_payload.temperature,
+        tfs=kai_payload.tfs,
+        top_p=kai_payload.top_p,
+        top_k=top_k,
+        top_a=kai_payload.top_a,
+        typical_p=kai_payload.typical,
+        stop=kai_payload.stop_sequence,
+        # ignore_eos=kai_payload.use_default_badwordsids, # TODO ban instead
+        max_tokens=kai_payload.max_length,
```

Comment:
I added the params according to the pr (henk717/KoboldAI#475 for reference)
> We could also add `frequency_penalty` here, since it's Kobold's alt repetition penalty.
The KAI wiki associates repetition penalty with the multiplicative variant. Are you saying the additive `frequency_penalty` also corresponds to a KAI param?

Reply:
> Are you saying the additive `frequency_penalty` also corresponds to a KAI param?

Alt rep pen with rep pen slope=1 is exactly freq pen, with the caveat that KAI accepts the exp(freq_pen) as rep_pen is that case and applies a log before adding it in.